benchmarking/caching/utils.py

```diff
@@ -0,0 +1,111 @@
+import os
+import re
+import shutil
+import sys
+from os import path
+from typing import List, Dict, Tuple, Union
+
+from api.models.team_set import TeamSet
+from benchmarking.caching.simulation_cache import SimulationCache
+
+
+SimulationArtifact = Tuple[List[TeamSet], List[float]]
+
+
+def combine(cache_key: str):
+    """
+    Takes a cache key and checks for any fragments. Combines them into one cache
+    file and uses the most recent data for the metadata. Removes fragments after
+    combining.
+
+    Example:
+    cache_key="test/abc" will be in fragments
+    simulation_cache
+        /test
+            /abc
+                /fragment_0.json
+                /fragment_1.json
+                ...
+    """
+
+    cache_directory = path.abspath(
+        path.join(path.dirname(__file__), "..", "..", "simulation_cache")
+    )
+    cache_location = path.normpath(path.join(cache_directory, cache_key)).replace(
+        "\\", "/"
+    )
+
+    # If cache not found, return
+    if not path.exists(cache_location):
+        return
+
+    # Get initial list of files where the cache should be
+    fragments = []
+    for f in os.listdir(cache_location):
+        match = re.match(r"(fragment_\d+)\.json", f)
+        if match:
+            fragments.append(match.group(1))
+
+    fragment_data: List[Dict[str, any]] = []
+
+    for fragment in fragments:
+        cache = SimulationCache(cache_key + "/" + fragment)
+        if not cache.exists():
+            continue
+        metadata = cache.get_metadata()
+        team_sets, runtimes = cache.get_simulation_artifact()
+
+        # Check that expected keys exist
+        if (
+            not metadata
+            or not metadata.get("commit_hash")
+            or not metadata.get("timestamp")
+            or not team_sets
+            or not runtimes
+        ):
+            print(
+                f"[WARNING]: Cache fragment missing critical data, skipping {fragment}",
+                file=sys.stderr,
+            )
+            continue
+
+        # Commit hashes must match
+        if (
+            len(fragment_data) > 0
+            and metadata["commit_hash"] != fragment_data[0]["metadata"]["commit_hash"]
```

Comment:
On second thought, why only check at index 0 here? I'm not sure I understand the data structure but this check and the error message it throws don't seem to match

Reply:
Oh nevermind I think I get it, feels a little gross of a way to write it though, probably just a nit; but idk something about this logic feels off. 

I think it's the fact we assume the first fragment is the source of truth? Idk